src/Pim/Component/Catalog/Model/ProductValueCollection.php

```diff
@@ -0,0 +1,275 @@
+<?php
+
+namespace Pim\Component\Catalog\Model;
+
+/**
+ * Business collection to handle product values.
+ *
+ * The collection is indexed internally by attribute-channel-locale. The index could be for instance:
+ *      description-ecommerce-en_US     for a localizable and scopable attribute
+ *      name-<all_channels>-en_US       for a localizable attribute
+ *      price-ecommerce-<all_locales>   for a scopable attribute
+ *
+ * This collection also contains the list of attributes used in the collection. The attributes
+ * are indexed by attribute codes.
+ *
+ * @author    Julien Janvier <j.janvier@gmail.com>
+ * @copyright 2017 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class ProductValueCollection implements ProductValueCollectionInterface
+{
+    /** @var ProductValueInterface[] */
+    private $values;
+
+    /** @var AttributeInterface[] */
+    private $attributes;
+
+    /**
+     * @param ProductValueInterface[] $values
+     */
+    public function __construct(array $values = array())
+    {
+        $this->values = [];
+        $this->attributes = [];
+
+        foreach ($values as $value) {
+            $this->add($value);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function toArray()
+    {
+        return $this->values;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function first()
+    {
+        return reset($this->values);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function last()
+    {
+        return end($this->values);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function key()
+    {
+        return key($this->values);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function next()
+    {
+        return next($this->values);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function current()
+    {
+        return current($this->values);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function removeKey($key)
+    {
+        if ( ! isset($this->values[$key]) && ! array_key_exists($key, $this->values)) {
+            return null;
+        }
+
+        $removed = $this->values[$key];
+        unset($this->values[$key]);
+
+        $this->reIndexAllAttributes();
+
+        return $removed;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function remove(ProductValueInterface $value)
+    {
+        $key = array_search($value, $this->values, true);
+
+        if ($key === false) {
+            return false;
+        }
+
+        unset($this->values[$key]);
+
+        $this->reIndexAllAttributes();
+
+        return true;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function containsKey($key)
+    {
+        return isset($this->values[$key]) || array_key_exists($key, $this->values);
```

Comment:
Why do you need both tests? Why not just array_key_exists, as we only look for the key and don't care about the value?